tests_e2e.py

```diff
@@ -0,0 +1,603 @@
+        response = request('post', 'auth/login', {}, {'username': user['username'], 'password': user['password']})
+        cookie = AuthHelper.get_hedy_cookie(response['headers']['Set-Cookie'])
+
+        # The cookie value must be set to `hedy={{SESSION}};` so that it can be used as a Cookie header in subsequent requests
+        USERS[user['username']]['cookie'] = CONFIG['session']['cookie_name'] + '=' + cookie.value + ';'
+        return user
+
+    @staticmethod
+    def assert_user_is_logged(username):
+        AuthHelper.assert_user_exists(username)
+        return AuthHelper.login_user(username)
+
+    @staticmethod
+    def get_hedy_cookie(cookie_string):
+        cookie = SimpleCookie()
+        cookie.load(cookie_string)
+
+        for key, cookie in cookie.items():
+            if key == CONFIG['session']['cookie_name']:
+                return cookie
+
+# *** TESTS ***
+
+class TestAuth(unittest.TestCase):
+    def test_invalid_signups(self):
+        # GIVEN a valid username
+        username = AuthHelper.make_username()
+        # WHEN attempting signups with invalid bodies
+        invalid_bodies = [
+            '',
+           [],
+            {},
+            {'username': 1},
+            {'username': 'user@me', 'password': 'foobar', 'email': 'a@a.com'},
+            {'username:': 'user: me', 'password': 'foobar', 'email': 'a@a.co'},
+            {'username': 't'},
+            {'username': '    t    '},
+            {'username': username},
+            {'username': username, 'password': 1},
+            {'username': username, 'password': 'foo'},
+            {'username': username, 'password': 'foobar'},
+            {'username': username, 'password': 'foobar', 'email': 'me@something'},
+            {'username': username, 'password': 'foobar', 'email': 'me@something.com', 'prog_experience':[2]},
+            {'username': username, 'password': 'foobar', 'email': 'me@something.com', 'prog_experience': 'foo'},
+            {'username': username, 'password': 'foobar', 'email': 'me@something.com', 'experience_languages': 'python'}
+        ]
+        for invalid_body in invalid_bodies:
+            response = request('post', 'auth/signup', {}, invalid_body)
+            # THEN receive an invalid response code from the server
+            self.assertEqual(response['code'], 400)
+
+    def test_signup(self):
+        # GIVEN a valid username and signup body
+        username = AuthHelper.make_username()
+        body = {'username': username, 'email': username + '@hedy.com', 'password': 'foobar'}
+
+        # WHEN signing up a new user
+        response = request('post', 'auth/signup', {}, body)
+
+        # THEN receive an OK response code from the server
+        self.assertEqual(response['code'], 200)
+        # THEN receive a body containing a token
+        self.assertIsInstance(response['body'], dict)
+        self.assertIsInstance(response['body']['token'], str)
+
+        # THEN Store the user and its token for upcoming tests
```

Comment:
I think technically this is not a #THEN because it does not concern the testing. Mayvbe we can use #FINALLY here (or is there another standard @rix0rrr?)